### PR TITLE
Migrate some diagnostics in `rustc_resolve` to session diagnostic

### PR DESCRIPTION
**Pull request by JeanCASPAR**
Hello, I migrated some diagnostics in `rustc_resolve` to session diagnostic.

r? @davidtwco

### CODE REVIEW

File: compiler/rustc_resolve/src/diagnostics.rs [hunk 1]
```diff
@@ -330,20 +335,20 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
...
-                err.tool_only_span_suggestion(
-                    import.use_span_with_attributes,
-                    "remove unnecessary import",
-                    "",
-                    Applicability::MaybeIncorrect,
+                err.subdiagnostic(
+                    self.tcx.dcx(),
+                    errors::ToolOnlyRemoveUnnecessaryImport {
+                        span: import.use_span_with_attributes,
+                    },
```
Comment (jieyouxu) [inline compiler/rustc_resolve/src/diagnostics.rs:347]: Hmm, is this conversion correct? docs for `tool_only_span_suggestion` says

> Adds a suggestion to the JSON output that will not be shown in the CLI.

Which I assume has different behavior from simply `err.subdiagnostic()` here?

File: compiler/rustc_resolve/src/diagnostics.rs [hunk 2]
```diff
@@ -460,22 +464,20 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         // previous imports.
         if found_closing_brace {
             if let Some(span) = extend_span_to_previous_binding(self.tcx.sess, span) {
-                err.tool_only_span_suggestion(span, message, "", Applicability::MaybeIncorrect);
+                err.subdiagnostic(self.dcx(), errors::ToolOnlyRemoveUnnecessaryImport { span });
```
Comment (jieyouxu) [inline compiler/rustc_resolve/src/diagnostics.rs:467]: And this

**Review by jieyouxu — COMMENTED** [review]
Looks good to me in general, ~~just two places where [`tool_only_span_suggestion`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/diagnostic/struct.Diag.html#method.tool_only_span_suggestion) was converted into `subdiagnostic()` which I believe changes the diagnostic output?~~ EDIT: `style = "tool-only"` was specified.

File: compiler/rustc_resolve/src/diagnostics.rs (hunk 1 above)
Comment (jieyouxu) [reply compiler/rustc_resolve/src/diagnostics.rs:347]: The impl for `tool_only_span_suggestion` is

```rs
#[rustc_lint_diagnostics]
pub fn tool_only_span_suggestion(
    &mut self,
    sp: Span,
    msg: impl Into<SubdiagMessage>,
    suggestion: impl ToString,
    applicability: Applicability,
) -> &mut Self {
    self.span_suggestion_with_style(
        sp,
        msg,
        suggestion,
        applicability,
        SuggestionStyle::CompletelyHidden,
    );
    self
}
```

Note the `SuggestionStyle::CompletelyHidden`. I wonder if you'll need a `err.tool_only_subdiagnostic()`?

Comment (JeanCASPAR) [reply compiler/rustc_resolve/src/diagnostics.rs:347]: `ToolOnlyRemoveUnnecessaryImport` has the following attribute :
```rust
#[suggestion(
    resolve_remove_unnecessary_import,
    code = "",
    applicability = "maybe-incorrect",
    style = "tool-only"
)]```
which should not show the suggestion in the CLI.

File: compiler/rustc_resolve/src/diagnostics.rs (hunk 2 above)
Comment (JeanCASPAR) [reply compiler/rustc_resolve/src/diagnostics.rs:467]: Same explanation.

File: compiler/rustc_resolve/src/diagnostics.rs (hunk 1 above)
Comment (jieyouxu) [reply compiler/rustc_resolve/src/diagnostics.rs:347]: Ah my bad, I completely overlooked that singular attribute (I was constantly jumping back and forth between the derives and the diagnostic construction sites lol).

**Comment by jieyouxu** [comment]
The conversions seem correct to me. Since this PR doesn't touch diagnostic infra, I'm going to steal the review and approve it.

Thanks for the PR!
r? jieyouxu
@bors r+ rollup

**Pull request merged**